tests/test_pytypes.cpp

```diff
@@ -126,6 +126,10 @@ TEST_SUBMODULE(pytypes, m) {
     m.def("tuple_ssize_t", []() { return py::tuple{(py::ssize_t) 0}; });
     m.def("tuple_size_t", []() { return py::tuple{(py::size_t) 0}; });
     m.def("get_tuple", []() { return py::make_tuple(42, py::none(), "spam"); });
+    m.def("access_tuple_with_int_index", []() {
+        py::object tpl = py::make_tuple(1, 2);
+        return tpl[1];
```

Comment:
int is implicitly converted to size_t first, I believe so it uses the tuple_accessor, not the item_accessor.

Reply:
Okay so it looks this test case itself is bugged and makes pypy complain.

Reply:
@woodychow Would you please add a dict case that actually tests the broken behavior? Like the dict example given in the issue?

Reply:
@Skylion007 for this particular case py::tuple could be used but what about multi-dimensional tuple/list?

Reply:
this test would be more proper here

```
    m.def("access_tuple_with_int_index", []() {
        py::object tpl = py::make_tuple(py::make_tuple(1, 2), py::make_tuple(3, 4));
        return tpl[0][1];
    });
```

this `tpl[0][1]` cannot be compiled on stock pybind11 even if tpl is declared as py::tuple.

Reply:
actually we need both single dimension case and multidimension cases...

Reply:
dict test https://github.com/cielavenir/pybind11/commit/fd82f0ab118f9e42d53abf9d6ba3cca6bb089670

somehow this one works fine, why tuple is broken like this...